Branching Round 1!

i2amroy opened the pull request:
A first stab at branching. Works in it's own right, though it may take a bit of finagling to get it actually working in the context of the whole. Currently utilizes all relative paths internally, though it should be easy to add an absolute caster pass method at the end for the rest of the program if that's the direction that that part still is utilizing at the end. Handles relative symlink generation within a file run, so it already handles the idea of allowing users to see all output for that particular run (even the reused part) in the single folder.

Things that this does not do:
* Handle generic argument paths - Right now it's hardcoded to look for 'infile' and 'outfile'; with some more finagling based on whether it is a top level or dependent module. Realistically we will need to add some sort of flags (path:true? See #46) to allow for determining when an input argument is a path and when said path is static rather than hardcoding to infile/outfile.

How to test:
1) Uncomment the test code at the bottom of the file and run
2) Place text files at locations actually running the pipeline would generate them at/pull them from (i.e., in.txt in OrchardTestStruct and out.txt/out2.txt/out3.txt in the newly created OrchardTestStruct/1 directory). You don't necessarily have to do this, but all the created links will be broken if those files are missing.
3) Edit the digit parameter in the config file.
4) Rerun the test code and you should see a new '2' directory created with symlinks to all files that will be reused in that particular run present.

jakereps commented:
Nice, thanks @i2amroy! I only had time right now to test run it and had one error pop up. It crashes when trying to pass a `pathlib.Path` object into `open`. Wrapping a `pathlib.Path` in `str` will give it's actual path as a plain string, or `pathlib.Path` objects allow the use of a native `.open()` method so it could be swapped to `work_path.open()` (line 196 and 213). After fixing 196 locally, it failed on 213 meaning it made the second branch folder, but didn't update the log, so after fixing 213 and rerunning it threw a "Folder already exists" error when trying to make 2 again, might be something we'd want to catch.

Will go through the actual code review in-depth later this weekend.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+                return False
+        if 'name' not in arg:
+            continue
+        if arg['name'] == name:
```
`if arg.get('name') == name:` will allow dropping line 46-47.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+        if 'arguments' not in module_2:
+            return False
+        for arg in module_1['arguments']:
+            argname = list(arg.keys())[0]
```
Dictionaries in Python are unordered. It's not guaranteed that `[0]` will be the key what's wanted.

i2amroy replied:
Each argument dictionary contains, and will only ever contain a single entry. As such I'm always guaranteed to get the key I want, because that's the only key that exists in the given dictionary (and if the config file is changed from that then all of this parsing is going to break anyways).

If you've got a better way to efficiently get the sole key from a dictionary, find the dictionary with a sole key that matches from a list of dict's, and compare the values for each key for equality then I'm all ears.

jakereps replied:
Ah ic, it wasn't clear that this was checking the config file and the `list()[0]` made it seem like it was a long list. If it's always (assumed) to be only one (key, value) pair, then the following will unpack it:

```python
argname, = arg # or arg.keys() to be visually explicit
match, = filter(lambda x: argname in x, module_2['arguments']) # again, or x.keys() to be visually specific
```

or this should do everything at once:

```python
mod1 = sorted(module_1['arguments'], key=lambda x: list(x))
mod2 = sorted(module_2['arguments'], key=lambda x: list(x))
for arg1, arg2 in zip(mod1, mod2):
    if list(arg1) != list(arg2) or list(arg1.values()) != list(arg2.values()):
        return False
```

jakereps replied:
Now that I think of it though, `zip` only joins up to the length of the shortest item, so it could false pass. That second snippet would probably need a `if len(mod1) != len(mod2): return False` before the loop.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+            if not is_branching_arg(link_module, argname):
+                continue
+            # Get the matching dictionary entry in the second config structure
+            match = next((item for item in module_2['arguments'] if list(item.keys())[0] == argname), None)
```
Same as above, no guarantee that `[0]` what is wanted.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+            # If we didn't find a match then they obviously don't match
+            if match is None:
+                return False
+            if arg[argname] != match[argname]:
```
Can drop a check with `if match is None or arg[argname] != match[argname]:`

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+            return False
+        # Check the normal optionals
+        for arg in module_1['optionals']:
+            argname = list(arg.keys())[0]
```
Same issue with dictionary ordering.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+            if not is_branching_in_list(link_module['optionals'], argname):
+                continue
+            # Get the matching dictionary entry in the second config structure
+            match = next((item for item in module_2['optionals'] if list(item.keys())[0] == argname), None)
```
Same as above, RE: dictionary ordering. These might be able to be turned into one function, just noticed how similar they are.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+    ret = split_file[0]
+    for step in split_file[1:]:
+        ret += '/' + step
+    return ret
```
This function can be reduced to just `os.path.join(pathstring, str(branchnum))`

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+
+def get_filename(pathstring):
+    split_file = pathstring.split("/")
+    return split_file[-1]
```
`os.path.basename(pathstring)`

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+def branching(config_file_data, link_file_data, workspace_path):
+    dep_struct = build_dep_struct(link_file_data)
+
+    work_path = Path(workspace_path + "/branchlog.yaml")
```
Not sure how much we care about cross-platform, but `work_path = Path(workspace_path) / 'branchlog.yaml'` will remove the hardcoded `/` from the path.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+
+    work_path = Path(workspace_path + "/branchlog.yaml")
+    if not work_path.is_file():
+        logfile = open(work_path, 'w+')
```
Best practice is to open a file within a context manager, so that it is auto-closed once the functionality within it is ran.

```python
with work_path.open('w+') as fh:
    ...
    yaml.safe_dump(logdata, fh, default_flow_style=False)
```

Or make sure to close the file at the end of the function.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+
+    best_branch = 0
+    branch_matching = []
+    logs = yaml.load(open(work_path, 'r').read())
```
Same as above.

```python
with work_path.open() as fh:
    logs = yaml.load(fh.read())
```

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+    while new_branch in logs:
+        new_branch += 1
+
+    os.makedirs(workspace_path + "/" + str(new_branch))
```
```python
os.makedirs(os.path.join(workspace_path, str(new_branch)))
```

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+    for module_name in branch_matching:
+        module = get_module(config_file_data, module_name)
+        if not module_in_top_level(dep_struct, module_name):
+            inpath = workspace_path + '/' + get_path(module, True)
```
`os.path.join`

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+
+    os.makedirs(workspace_path + "/" + str(new_branch))
+    logs[new_branch] = config_file_data
+    logfile = open(work_path, 'w')
```
Context manager, or close.

jakereps commented on orchard/modules/_branch.py:
```diff
@@ -6,6 +6,228 @@
+            newinpath = insert_branch_into_path(inpath, new_branch)
+            if not os.path.exists(newinpath):
+                os.symlink(os.path.relpath(oldinpath, os.path.dirname(newinpath)), newinpath)
+        outpath = workspace_path + '/' + get_path(module, False)
```
`os.path.join`

i2amroy commented:
Addressed all of the commented issues except the dictionary[0] one which I responded to above. To address your issue of crashing, what version of Python are you running it in? Because it's working just fine in my Python 3.6 (which looks like it made some changes to the ability of certain functions to handle "path-like structures", and as such could be the issue you are running into).

jakereps commented:
LGTM 👍  🎉  🎊
